hotshot-query-service/src/data_source/storage/sql.rs

```diff
@@ -1495,4 +1620,123 @@ mod test {
+                builder_commitment,
+                metadata,
+            );
+
+            block_header.block_number = i;
+
+            let null_quorum_data = QuorumData {
+                leaf_commit: Commitment::<Leaf<MockTypes>>::default_commitment_no_preimage(),
+            };
+
+            let mut qc = QuorumCertificate::new(
+                null_quorum_data.clone(),
+                null_quorum_data.commit(),
+                view,
+                None,
+                std::marker::PhantomData,
+            );
+
+            let quorum_proposal = QuorumProposal {
+                block_header,
+                view_number: view,
+                justify_qc: qc.clone(),
+                upgrade_certificate: None,
+                proposal_certificate: None,
+            };
+
+            let mut leaf = Leaf::from_quorum_proposal(&quorum_proposal);
+            leaf.fill_block_payload(payload, 4).unwrap();
+            qc.data.leaf_commit = <Leaf<MockTypes> as Committable>::commit(&leaf);
+
+            let height = leaf.height() as i64;
+            let hash = <Leaf<_> as Committable>::commit(&leaf).to_string();
+            let header = leaf.block_header();
+
+            let header_json = serde_json::to_value(header)
+                .context("failed to serialize header")
+                .unwrap();
+
+            let payload_commitment =
+                <MockHeader as BlockHeader<MockTypes>>::payload_commitment(header);
+            let mut tx = storage.write().await.unwrap();
+
+            tx.upsert(
+                "header",
+                ["height", "hash", "payload_hash", "data", "timestamp"],
+                ["height"],
+                [(
+                    height,
+                    leaf.block_header().commit().to_string(),
+                    payload_commitment.to_string(),
+                    header_json,
+                    leaf.block_header().timestamp() as i64,
+                )],
+            )
+            .await
+            .unwrap();
+
+            let leaf_json = serde_json::to_value(leaf.clone()).expect("failed to serialize leaf");
+            let qc_json = serde_json::to_value(qc).expect("failed to serialize QC");
+            tx.upsert(
+                "leaf",
+                ["height", "hash", "block_hash", "leaf", "qc"],
+                ["height"],
+                [(
+                    height,
+                    hash,
+                    header.commit().to_string(),
+                    leaf_json,
+                    qc_json,
+                )],
+            )
+            .await
+            .unwrap();
+            tx.commit().await.unwrap();
+        }
+
+        storage
+            .migrate_types::<MockTypes>()
+            .await
+            .expect("failed to migrate");
+        let mut tx = storage.read().await.unwrap();
+        let (count,) = query_as::<(i64,)>("SELECT COUNT(*) from leaf2")
+            .fetch_one(tx.as_mut())
+            .await
+            .unwrap();
```

Comment:
Instead of checking the number of rows it would be better to check that the data we migrated is as expected.

Reply:
in my todo list

Reply:
https://github.com/EspressoSystems/espresso-sequencer/issues/2710